packages/react-native/ReactCommon/react/nativemodule/core/platform/ios/ReactCommon/RCTTurboModule.mm

```diff
@@ -340,12 +387,51 @@ id convertJSIValueToObjCObject(jsi::Runtime &runtime, const jsi::Value &value, s
       TurboModulePerfLogger::asyncMethodCallExecutionStart(moduleName, methodNameStr.c_str(), asyncCallCounter);
     }
 
-    @try {
-      [inv invokeWithTarget:strongModule];
-    } @catch (NSException *exception) {
-      throw convertNSExceptionToJSError(runtime, exception);
-    } @finally {
-      [retainedObjectsForInvocation removeAllObjects];
+    NSError *caughtException = nil;
+    BOOL shouldCatchException = isSync || reject;
+    try {
+      @try {
+        [inv invokeWithTarget:strongModule];
+       } @catch (NSException *exception) {
+         caughtException = maybeCatchException(
+                     shouldCatchException, @{
+                       @"name": exception.name,
+                       NSLocalizedDescriptionKey: exception.reason,
+                       @"stackSymbols": exception.callStackSymbols,
+                       @"stackReturnAddresses": exception.callStackReturnAddresses,
+                     });
+       } @catch (NSError *error) {
+         caughtException = maybeCatchException(shouldCatchException, error);
+       } @catch (NSString *errorMessage) {
+         caughtException = maybeCatchException(shouldCatchException, @{
+                      NSLocalizedDescriptionKey: errorMessage,
+                     });
+       } @catch (id e) {
+         caughtException = maybeCatchException(shouldCatchException, @{
+                      NSLocalizedDescriptionKey: @"Unknown Objective-C Object thrown.",
+                     });
+       }
+    } catch (const std::exception &exception) {
+      caughtException = maybeCatchException(shouldCatchException, @{
+                    NSLocalizedDescriptionKey: [NSString stringWithUTF8String:exception.what()],
+                  });
+    } catch (const std::string &errorMessage) {
+      caughtException = maybeCatchException(shouldCatchException, @{
+                    NSLocalizedDescriptionKey: [NSString stringWithUTF8String:errorMessage.c_str()],
+                  });
+    } catch (...) {
+      caughtException = maybeCatchException(shouldCatchException, @{
+                    NSLocalizedDescriptionKey: @"Unknown C++ exception thrown.",
+                  });
+    }
+    [retainedObjectsForInvocation removeAllObjects];
```

Comment:
This still needs to go in a @finally, otherwise, we'll leak `retainedObjectsForInvocation` when exceptions are thrown.